[Feature][MaterializedView] add AlterMaterializedViewStmt for MaterializedView

zhuxt2015 opened the pull request:
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：
<!--
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #4782

## Problem Summary(Required) ：
<!-- (Please describe the changes you have made. In which scenarios will this bug be triggered and what measures have you taken to fix the bug?) -->

zhuxt2015 committed:
add AlterMaterializedViewStmt for MaterializedView

zhuxt2015 committed:
Alter MV use new Analyzer

zhuxt2015 committed:
fix

wanpengfei-git commented:
#### [FE PR Coverage check]

:disappointed: **fail** : 36 / 105 (34.29%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/alter/Alter.java|0|30|00.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/alter/AlterJobV2.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/connector/ConnectorMetadata.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/alter/RollupJobV2.java|0|5|00.00%|
|:large_blue_circle:|com/starrocks/alter/MaterializedViewHandler.java|5|27|18.52%|
|:large_blue_circle:|com/starrocks/qe/DdlExecutor.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|9|9|100.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AlterMaterializedViewStmt.java|9|9|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|10|10|100.00%|

zhuxt2015 committed:
fix

zhuxt2015 committed:
Merge branch 'main' of https://github.com/StarRocks/starrocks into alter_mv

zhuxt2015 committed:
fix

wanpengfei-git commented:
#### [FE PR Coverage check]

:disappointed: **fail** : 33 / 155 (21.29%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/alter/Alter.java|0|84|00.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/alter/AlterJobV2.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|0|6|00.00%|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|0|10|00.00%|
|:large_blue_circle:|com/starrocks/connector/ConnectorMetadata.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/alter/RollupJobV2.java|0|5|00.00%|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|2|5|40.00%|
|:large_blue_circle:|com/starrocks/qe/DdlExecutor.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|9|9|100.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AlterMaterializedViewStatement.java|9|9|100.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|10|10|100.00%|

ABingHuang commented on fe/fe-core/src/main/java/com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java:
```diff
@@ -263,6 +266,25 @@ public Void visitDropMaterializedViewStatement(DropMaterializedViewStmt stmt, Co
+                    throw new SemanticException("New table name can not equals to old name");
+                }
+            }
+            final RefreshSchemeDesc refreshSchemeDesc = statement.getRefreshSchemeDesc();
```
here should support modify time interval? it seems that RefreshSchemeDesc do not contain interval field

ABingHuang commented on fe/fe-core/src/main/java/com/starrocks/alter/Alter.java:
```diff
@@ -188,6 +193,128 @@ public void processDropMaterializedView(DropMaterializedViewStmt stmt) throws Dd
+            EditLog editLog = GlobalStateMgr.getCurrentState().getEditLog();
+            //rename materialized view
+            if (newMvName != null) {
+                if (oldMvName.equals(newMvName)) {
```
can use a function for each branch

ABingHuang commented on fe/fe-core/src/main/java/com/starrocks/alter/Alter.java:
```diff
@@ -188,6 +193,128 @@ public void processDropMaterializedView(DropMaterializedViewStmt stmt) throws Dd
+                newMvRefreshScheme.setType(MaterializedView.RefreshType.valueOf(refreshType));
+                newMvRefreshScheme.setLastRefreshTime(oldRefreshScheme.getLastRefreshTime());
+                materializedView.setRefreshScheme(newMvRefreshScheme);
+                editLog.logMvChangeRefreshScheme(materializedView);
```
should add a special log for this

ABingHuang commented on fe/fe-core/src/main/java/com/starrocks/alter/Alter.java:
```diff
@@ -188,6 +193,128 @@ public void processDropMaterializedView(DropMaterializedViewStmt stmt) throws Dd
+                materializedView.setName(newMvName);
+                db.dropTable(oldMvName);
+                db.createTable(materializedView);
+                editLog.logMvRename(materializedView);
```
should add a special log for this

ABingHuang commented on fe/fe-core/src/main/java/com/starrocks/alter/Alter.java:
```diff
@@ -188,6 +193,128 @@ public void processDropMaterializedView(DropMaterializedViewStmt stmt) throws Dd
+                materializedView.setRefreshScheme(newMvRefreshScheme);
+                editLog.logMvChangeRefreshScheme(materializedView);
+            } else {
+                throw new DdlException("materialized view no change");
```
```suggestion
                throw new DdlException("unsupported modification for materialized view");
```

zhuxt2015 committed:
fix

zhuxt2015 committed:
Merge branch 'main' of https://github.com/StarRocks/starrocks into alter_mv

# Conflicts:
#	fe/fe-core/src/main/java/com/starrocks/persist/OperationType.java

zhuxt2015 committed:
fix

zhuxt2015 committed:
fix

wanpengfei-git commented:
#### [FE PR Coverage check]

:disappointed: **fail** : 66 / 266 (24.81%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/alter/Alter.java|0|99|00.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|0|6|00.00%|
|:large_blue_circle:|com/starrocks/persist/RenameMaterializedViewLog.java|0|18|00.00%|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|0|10|00.00%|
|:large_blue_circle:|com/starrocks/persist/ChangeMaterializedViewRefreshSchemeLog.java|0|23|00.00%|
|:large_blue_circle:|com/starrocks/connector/ConnectorMetadata.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|8|00.00%|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|3|19|15.79%|
|:large_blue_circle:|com/starrocks/qe/DdlExecutor.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AsyncRefreshSchemeDesc.java|8|11|72.73%|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|20|25|80.00%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|22|25|88.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AlterMaterializedViewStatement.java|9|9|100.00%|
|:large_blue_circle:|com/starrocks/sql/ast/RefreshSchemeDesc.java|1|1|100.00%|

zhuxt2015 committed:
fix

zhuxt2015 committed:
remove no referenced import

wanpengfei-git commented:
#### [FE PR Coverage check]

:disappointed: **fail** : 66 / 266 (24.81%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/alter/Alter.java|0|103|00.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|0|6|00.00%|
|:large_blue_circle:|com/starrocks/persist/RenameMaterializedViewLog.java|0|17|00.00%|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|0|11|00.00%|
|:large_blue_circle:|com/starrocks/persist/ChangeMaterializedViewRefreshSchemeLog.java|0|21|00.00%|
|:large_blue_circle:|com/starrocks/connector/ConnectorMetadata.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|6|00.00%|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|3|19|15.79%|
|:large_blue_circle:|com/starrocks/qe/DdlExecutor.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AsyncRefreshSchemeDesc.java|8|11|72.73%|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|20|25|80.00%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|22|25|88.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AlterMaterializedViewStatement.java|9|9|100.00%|
|:large_blue_circle:|com/starrocks/sql/ast/RefreshSchemeDesc.java|1|1|100.00%|

Astralidea commented on fe/fe-core/src/main/java/com/starrocks/alter/Alter.java:
```diff
@@ -188,6 +199,157 @@ public void processDropMaterializedView(DropMaterializedViewStmt stmt) throws Dd
+                materializedView = (MaterializedView) table;
+            }
+            if (materializedView == null) {
+                for (Table t : db.getTables()) {
```
what's the different between traverse db.getTables() db.getTable(oldMvName);

Astralidea commented on fe/fe-core/src/main/java/com/starrocks/alter/Alter.java:
```diff
@@ -188,6 +199,157 @@ public void processDropMaterializedView(DropMaterializedViewStmt stmt) throws Dd
+            EditLog editLog = GlobalStateMgr.getCurrentState().getEditLog();
+            //rename materialized view
+            if (newMvName != null) {
+                processRenameMaterializedView(oldMvName, newMvName, db, materializedView, editLog);
```
editLog is from GlobalStateMgr and is not need to pass argument, when you use where you can get.

Astralidea commented on fe/fe-core/src/main/java/com/starrocks/alter/Alter.java:
```diff
@@ -188,6 +199,157 @@ public void processDropMaterializedView(DropMaterializedViewStmt stmt) throws Dd
+                processRenameMaterializedView(oldMvName, newMvName, db, materializedView, editLog);
+            } else if (refreshSchemeDesc != null) {
+                //change refresh scheme
+                processChangeRefreshScheme(refreshSchemeDesc, materializedView, editLog);
```
will alter both Name and refreshSchemaDesc one time?

zhuxt2015 replied:
According to the current design, alter both is not supported

Astralidea commented on fe/fe-core/src/main/java/com/starrocks/alter/Alter.java:
```diff
@@ -188,6 +199,157 @@ public void processDropMaterializedView(DropMaterializedViewStmt stmt) throws Dd
+        TimestampArithmeticExpr.TimeUnit timeUnit = null;
+        if (refreshSchemeDesc instanceof AsyncRefreshSchemeDesc) {
+            newMvRefreshScheme.setAsyncRefreshContext(oldRefreshScheme.getAsyncRefreshContext());
+            startTime = ((AsyncRefreshSchemeDesc) refreshSchemeDesc).getStartTime();
```
better to use
```
AsyncRefreshSchemeDesc asyncRefreshSchemeDesc = (AsyncRefreshSchemeDesc) refreshSchemeDesc;
newMvRefreshScheme.setStartTime(asyncRefreshSchemeDesc.getStartTime());
```

Astralidea commented on fe/fe-core/src/main/java/com/starrocks/persist/ChangeMaterializedViewRefreshSchemeLog.java:
```diff
@@ -0,0 +1,82 @@
+    private RefreshType refreshType;
+
+    @SerializedName(value = "startTime")
+    private LocalDateTime startTime;
```
why not use long to persist Time?

Astralidea commented on fe/fe-core/src/main/java/com/starrocks/alter/Alter.java:
```diff
@@ -188,6 +199,157 @@ public void processDropMaterializedView(DropMaterializedViewStmt stmt) throws Dd
+        materializedView.setRefreshScheme(newMvRefreshScheme);
+        final ChangeMaterializedViewRefreshSchemeLog log =
+                new ChangeMaterializedViewRefreshSchemeLog(materializedView.getId(), materializedView.getDbId(),
+                        newRefreshType, startTime, step, timeUnit);
```
why not pass argument class AsyncRefreshSchemeDesc?

ABingHuang commented on fe/fe-core/src/main/java/com/starrocks/alter/Alter.java:
```diff
@@ -188,6 +199,157 @@ public void processDropMaterializedView(DropMaterializedViewStmt stmt) throws Dd
+            throw new DdlException("Materialized view [" + newMvName + "] is already used");
+        }
+        materializedView.setName(newMvName);
+        db.dropTable(oldMvName);
```
just rename the materialized view. why should drop mv and recreate another?

ABingHuang commented on fe/fe-core/src/main/java/com/starrocks/alter/Alter.java:
```diff
@@ -188,6 +199,157 @@ public void processDropMaterializedView(DropMaterializedViewStmt stmt) throws Dd
+        try {
+            MaterializedView oldMaterializedView = (MaterializedView) db.getTable(materializedViewId);
+            String oldMaterializedViewName = oldMaterializedView.getName();
+            db.dropTable(oldMaterializedViewName);
```
no need to drop old and recreate new

ABingHuang commented on fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java:
```diff
@@ -59,16 +56,33 @@ public static class MvRefreshScheme {
         private long lastRefreshTime;
 
+        @SerializedName(value = "starTime")
+        private LocalDateTime startTime;
```
better to move these properties into AsyncRefreshContext.

zhuxt2015 committed:
fix

wanpengfei-git commented:
#### [FE PR Coverage check]

:disappointed: **fail** : 66 / 238 (27.73%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/alter/Alter.java|0|86|00.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|0|6|00.00%|
|:large_blue_circle:|com/starrocks/persist/RenameMaterializedViewLog.java|0|15|00.00%|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|0|11|00.00%|
|:large_blue_circle:|com/starrocks/persist/ChangeMaterializedViewRefreshSchemeLog.java|0|15|00.00%|
|:large_blue_circle:|com/starrocks/connector/ConnectorMetadata.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|6|00.00%|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|3|16|18.75%|
|:large_blue_circle:|com/starrocks/qe/DdlExecutor.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AsyncRefreshSchemeDesc.java|8|11|72.73%|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|20|25|80.00%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|22|25|88.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AlterMaterializedViewStatement.java|9|9|100.00%|
|:large_blue_circle:|com/starrocks/sql/ast/RefreshSchemeDesc.java|1|1|100.00%|

ABingHuang commented on fe/fe-core/src/main/java/com/starrocks/alter/Alter.java:
```diff
@@ -188,6 +198,140 @@ public void processDropMaterializedView(DropMaterializedViewStmt stmt) throws Dd
+            throw new DdlException("Materialized view [" + newMvName + "] is already used");
+        }
+        materializedView.setName(newMvName);
+        db.createTable(materializedView);
```
should drop and recreate table like OlapTable
it is my fault.

zhuxt2015 replied:
That's all right.

ABingHuang commented on fe/fe-core/src/main/java/com/starrocks/alter/Alter.java:
```diff
@@ -188,6 +198,140 @@ public void processDropMaterializedView(DropMaterializedViewStmt stmt) throws Dd
+        try {
+            MaterializedView oldMaterializedView = (MaterializedView) db.getTable(materializedViewId);
+            oldMaterializedView.setName(newMaterializedViewName);
+            db.createTable(oldMaterializedView);
```
should drop and recreate table like OlapTable

ABingHuang commented on fe/fe-core/src/main/java/com/starrocks/sql/ast/AsyncRefreshSchemeDesc.java:
```diff
@@ -10,20 +12,53 @@ public class AsyncRefreshSchemeDesc extends RefreshSchemeDesc {
-    public AsyncRefreshSchemeDesc(LocalDateTime startTime, IntervalLiteral intervalLiteral) {
+    private TimestampArithmeticExpr.TimeUnit timeUnit;
+
+    private final ImmutableSet<TimestampArithmeticExpr.TimeUnit> supportedTimeUnitType = ImmutableSet.of(
```
it seems that it has been defined in  RefreshSchemeDesc

zhuxt2015 committed:
fix

wanpengfei-git commented:
#### [FE PR Coverage check]

:disappointed: **fail** : 65 / 234 (27.78%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/alter/Alter.java|0|83|00.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|0|6|00.00%|
|:large_blue_circle:|com/starrocks/persist/RenameMaterializedViewLog.java|0|15|00.00%|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|0|11|00.00%|
|:large_blue_circle:|com/starrocks/persist/ChangeMaterializedViewRefreshSchemeLog.java|0|15|00.00%|
|:large_blue_circle:|com/starrocks/connector/ConnectorMetadata.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|6|00.00%|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|3|16|18.75%|
|:large_blue_circle:|com/starrocks/qe/DdlExecutor.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AsyncRefreshSchemeDesc.java|8|11|72.73%|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|20|25|80.00%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|22|25|88.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AlterMaterializedViewStatement.java|9|9|100.00%|

zhuxt2015 committed:
Merge branch 'main' of https://github.com/StarRocks/starrocks into alter_mv

zhuxt2015 commented:
run starrocks_fe_unittest

zhuxt2015 committed:
Merge branch 'main' of https://github.com/StarRocks/starrocks into alter_mv

wanpengfei-git commented:
#### [FE PR Coverage check]

:disappointed: **fail** : 65 / 234 (27.78%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/alter/Alter.java|0|83|00.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|0|6|00.00%|
|:large_blue_circle:|com/starrocks/persist/RenameMaterializedViewLog.java|0|15|00.00%|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|0|11|00.00%|
|:large_blue_circle:|com/starrocks/persist/ChangeMaterializedViewRefreshSchemeLog.java|0|15|00.00%|
|:large_blue_circle:|com/starrocks/connector/ConnectorMetadata.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|6|00.00%|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|3|16|18.75%|
|:large_blue_circle:|com/starrocks/qe/DdlExecutor.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AsyncRefreshSchemeDesc.java|8|11|72.73%|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|20|25|80.00%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|22|25|88.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AlterMaterializedViewStatement.java|9|9|100.00%|

ABingHuang commented on fe/fe-core/src/main/java/com/starrocks/alter/Alter.java:
```diff
@@ -188,6 +198,135 @@ public void processDropMaterializedView(DropMaterializedViewStmt stmt) throws Dd
+            // check materialized view state
+            if (materializedView.getState() != OlapTableState.NORMAL) {
+                throw new DdlException("Materialized view [" + materializedView.getName() + "]'s state is not NORMAL. "
+                        + "Do not allow doing ALTER ops");
```
```suggestion
                        + "Do not allow to do ALTER ops");
```

ABingHuang reviewed (commented):
LGTM

zhuxt2015 committed:
Update fe/fe-core/src/main/java/com/starrocks/alter/Alter.java

Co-authored-by: ABing <101158374+ABingHuang@users.noreply.github.com>

Astralidea commented on fe/fe-core/src/main/java/com/starrocks/alter/Alter.java:
```diff
@@ -188,6 +198,135 @@ public void processDropMaterializedView(DropMaterializedViewStmt stmt) throws Dd
+            throw new DdlException("unsupported change to SYNC refresh type");
+        }
+
+        if (refreshSchemeDesc instanceof AsyncRefreshSchemeDesc) {
```
better to use 
if (refreshSchemeDesc instanceof AsyncRefreshSchemeDesc) { xxx }
else if if (refreshSchemeDesc instanceof SyncRefreshSchemeDesc) { xxx }

Astralidea commented on fe/fe-core/src/main/java/com/starrocks/alter/Alter.java:
```diff
@@ -188,6 +198,135 @@ public void processDropMaterializedView(DropMaterializedViewStmt stmt) throws Dd
+        final RefreshType newRefreshType = RefreshType.valueOf(refreshType);
+        refreshScheme.setType(newRefreshType);
+        final ChangeMaterializedViewRefreshSchemeLog log = new ChangeMaterializedViewRefreshSchemeLog(materializedView);
+        EditLog editLog = GlobalStateMgr.getCurrentState().getEditLog();
```
```suggestion
       GlobalStateMgr.getCurrentState().getEditLog().logMvChangeRefreshScheme(log);
```

Astralidea commented on fe/fe-core/src/main/java/com/starrocks/alter/Alter.java:
```diff
@@ -188,6 +198,135 @@ public void processDropMaterializedView(DropMaterializedViewStmt stmt) throws Dd
+        db.createTable(materializedView);
+        final RenameMaterializedViewLog renameMaterializedViewLog =
+                new RenameMaterializedViewLog(materializedView, oldMvName, newMvName);
+        EditLog editLog = GlobalStateMgr.getCurrentState().getEditLog();
```
```suggestion
        GlobalStateMgr.getCurrentState().getEditLog().logMvRename(renameMaterializedViewLog);
```

kangkaisen replied:
I think we could put the EditLog.logxxx method to the outside of DB lock. because the log operation maybe need a long time.

Astralidea commented on fe/fe-core/src/main/java/com/starrocks/alter/Alter.java:
```diff
@@ -188,6 +198,135 @@ public void processDropMaterializedView(DropMaterializedViewStmt stmt) throws Dd
+            AsyncRefreshSchemeDesc asyncRefreshSchemeDesc = (AsyncRefreshSchemeDesc) refreshSchemeDesc;
+            final MaterializedView.AsyncRefreshContext asyncRefreshContext = refreshScheme.getAsyncRefreshContext();
+            asyncRefreshContext.setStartTime(Utils.getLongFromDateTime(asyncRefreshSchemeDesc.getStartTime()));
+            asyncRefreshContext.setStep(asyncRefreshSchemeDesc.getStep());
```
maybe here should be check Step > 0 ?

zhuxt2015 commented on fe/fe-core/src/main/java/com/starrocks/persist/OperationType.java:
```diff
@@ -225,4 +225,6 @@ public class OperationType {
     public static final short OP_UPDATE_TASK_RUN = 10082;
     public static final short OP_DROP_TASK_RUNS = 10083;
+    public static final short OP_RENAME_MATERIALIZED_VIEW = 10091;
+    public static final short OP_CHANGE_MATERIALIZED_VIEW_REFRESH_SCHEME = 10092;
```
Refer to table's design, it is better to separate the types of operations

zhuxt2015 committed:
fix

zhuxt2015 committed:
Merge branch 'main' of https://github.com/StarRocks/starrocks into alter_mv

zhuxt2015 committed:
Merge remote-tracking branch 'origin/alter_mv' into alter_mv

wanpengfei-git commented:
#### [FE PR Coverage check]

:disappointed: **fail** : 68 / 236 (28.81%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/alter/Alter.java|0|80|00.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|0|6|00.00%|
|:large_blue_circle:|com/starrocks/persist/RenameMaterializedViewLog.java|0|15|00.00%|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|0|11|00.00%|
|:large_blue_circle:|com/starrocks/persist/ChangeMaterializedViewRefreshSchemeLog.java|0|15|00.00%|
|:large_blue_circle:|com/starrocks/connector/ConnectorMetadata.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|6|00.00%|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|3|16|18.75%|
|:large_blue_circle:|com/starrocks/qe/DdlExecutor.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AsyncRefreshSchemeDesc.java|8|11|72.73%|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|22|28|78.57%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|23|27|85.19%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AlterMaterializedViewStatement.java|9|9|100.00%|

zhuxt2015 committed:
Merge branch 'main' of https://github.com/StarRocks/starrocks into alter_mv

 Conflicts:
	fe/fe-core/src/main/java/com/starrocks/sql/ast/AstVisitor.java

ABingHuang commented on fe/fe-core/src/main/java/com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java:
```diff
@@ -265,6 +271,50 @@ public Void visitDropMaterializedViewStatement(DropMaterializedViewStmt stmt, Co
+                        throw new SemanticException("please input interval clause");
+                    }
+                    if (startTime != null) {
+                        if (startTime.isBefore(LocalDateTime.now())) {
```
```suggestion
                        if (startTime != null && startTime.isBefore(LocalDateTime.now())) {
```

ABingHuang commented on fe/fe-core/src/main/java/com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java:
```diff
@@ -265,6 +271,50 @@ public Void visitDropMaterializedViewStatement(DropMaterializedViewStmt stmt, Co
+                        async.setStartTime(LocalDateTime.now());
+                    }
+                    if (timeUnit != null) {
+                        if (!async.getSupportedTimeUnitType().contains(timeUnit)) {
```
```suggestion
                        if (timeUnit != null && !async.getSupportedTimeUnitType().contains(timeUnit)) {
```

zhuxt2015 committed:
Merge remote-tracking branch 'origin/alter_mv' into alter_mv

# Conflicts:
#	fe/fe-core/src/main/java/com/starrocks/persist/OperationType.java
#	fe/fe-core/src/main/java/com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java

zhuxt2015 committed:
fix

zhuxt2015 committed:
fix

zhuxt2015 commented:
run starrocks_fe_unittest

zhuxt2015 committed:
Merge branch 'main' of https://github.com/StarRocks/starrocks into alter_mv

wanpengfei-git commented:
#### [FE PR Coverage check]

:disappointed: **fail** : 70 / 237 (29.54%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/alter/Alter.java|0|81|00.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|0|6|00.00%|
|:large_blue_circle:|com/starrocks/persist/RenameMaterializedViewLog.java|0|15|00.00%|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|0|11|00.00%|
|:large_blue_circle:|com/starrocks/persist/ChangeMaterializedViewRefreshSchemeLog.java|0|15|00.00%|
|:large_blue_circle:|com/starrocks/connector/ConnectorMetadata.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|6|00.00%|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|3|16|18.75%|
|:large_blue_circle:|com/starrocks/qe/DdlExecutor.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AsyncRefreshSchemeDesc.java|8|11|72.73%|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|23|28|82.14%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|24|27|88.89%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AlterMaterializedViewStatement.java|9|9|100.00%|

Astralidea commented on fe/fe-core/src/main/java/com/starrocks/persist/ChangeMaterializedViewRefreshSchemeLog.java:
```diff
@@ -0,0 +1,61 @@
+
+public class ChangeMaterializedViewRefreshSchemeLog implements Writable {
+    @SerializedName(value = "MaterializedViewId")
+    private long id;
```
same here

ABingHuang reviewed (commented):
LGTM

zhuxt2015 committed:
fix

ABingHuang commented:
run starrocks_fe_unittest

wangruin commented:
run starrocks_fe_unittest

ABingHuang commented:
run starrocks_fe_unittest

HangyuanLiu commented on fe/fe-core/src/main/java/com/starrocks/sql/ast/AsyncRefreshSchemeDesc.java:
```diff
@@ -10,20 +12,53 @@ public class AsyncRefreshSchemeDesc extends RefreshSchemeDesc {
+    private long step;
 
-    public AsyncRefreshSchemeDesc(LocalDateTime startTime, IntervalLiteral intervalLiteral) {
+    private TimestampArithmeticExpr.TimeUnit timeUnit;
```
TimestampArithmeticExpr has nothing to do with here, why only use TimestampArithmeticExpr? Why not use interval?

HangyuanLiu commented on fe/fe-core/src/main/java/com/starrocks/alter/Alter.java:
```diff
@@ -188,6 +198,128 @@ public void processDropMaterializedView(DropMaterializedViewStmt stmt) throws Dd
+    private void processRenameMaterializedView(String oldMvName, String newMvName, Database db,
+                                               MaterializedView materializedView) throws DdlException {
+        if (oldMvName.equals(newMvName)) {
+            throw new DdlException("Same materialized view name");
```
The logic that has been verified in the Analyzer does not need to be repeatedly verified

HangyuanLiu commented on fe/fe-core/src/main/java/com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java:
```diff
@@ -265,6 +271,50 @@ public Void visitDropMaterializedViewStatement(DropMaterializedViewStmt stmt, Co
         }
+
+        @Override
+        public Void visitAlterMaterializedViewStatement(AlterMaterializedViewStatement statement,
```
The syntax error message will be returned to the user, and we should give some more specific error messages.

HangyuanLiu reviewed (changes requested):
There is some content to discuss

zhuxt2015 committed:
Merge branch 'main' of https://github.com/StarRocks/starrocks into alter_mv

# Conflicts:
#	fe/fe-core/src/main/java/com/starrocks/persist/EditLog.java
#	fe/fe-core/src/main/java/com/starrocks/persist/OperationType.java
#	fe/fe-core/src/main/java/com/starrocks/sql/StatementPlanner.java
#	fe/fe-core/src/main/java/com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java

HangyuanLiu committed:
Fix sql_full_groupby mode mistake rewrite column in aggregation function bug (#7049)

meegoo committed:
[BugFix] Fix commit timeout invalid (#7051)

adzfolc committed:
[Feature]Be support customize java opts (#6976)

xlwh committed:
Fix compile issue in ARM X64 and with Ninja (#6878)

Linkerist committed:
conditional loop not for loop with break to improve readability. (#7027)

imay committed:
[Refactor] Split column_viewer.h (#7058)

rickif committed:
[BugFix] Support transferring json using HTTP chunk (#6972)

This PR is the last part of the fix of #5515, including the following modifications:
1. building JSON buffer in the handler of stream load to support chunked JSON HTTP body
2. removing `Status MessageBodySink::append(const ByteBufferPtr& buf)` to reduce memory copy

sduzh committed:
[Refactor] Move impl of AsyncDeltaWriterExecutor to util/bthreads directory (#7038)

Youngwb committed:
[BugFix] Fix Enforcer record error required property (#7014)

decster committed:
[Enhancement] Report tablet version after publish version task finish (#6537)

Currently, BE's publish_version task only reports whether it is successful or not but does not contain which version the replica has and which versions are missing. This will limit FE's ability to publish txn or schedule clone tasks.

For example, if a replica has version 1 2 4 5, and publish_version 3 succeeds, the version should grow to 5, and txns waiting for version 4, 5 to be published can proceed.

This PR also does a major refactor to the PublishVersionTask execution-related code.

ZiheLiu committed:
[Bugfix] process returned status correctly (#7040)

murphyatwork committed:
Remove unused in iterator predicate (#3615)

liuyehcf committed:
[Enhance] add PENDING_FINISH_TIME in profile (#7074)

Seaven committed:
[BugFix] Fix throw ConcurrentModificationException use partition range (#7076)

Astralidea committed:
Batch create partition support dynamic_partition.prefix (#7016)

consider into dynamic_partition.prefix, and I think this only should support for date partition. because dynamic partition only create by date.

stdpain committed:
[Bugfix] fix call JNI in bthread in Java UDF (#7092)

stephen-shelby committed:
update name and type when showing catalogs (#7093)

Youngwb committed:
[BugFix] Remove wrong preconditions in join reorder (#7099)

aaawuanjun committed:
[BugFix]  fix issue 6732 about bitmap_max function data overflow (#6768)

sduzh committed:
Refine impl of MemTable (#7094)

sduzh committed:
[Lake] Add AsyncDeltaWriter (#7091)

Seaven committed:
[Bug] Fix in-subquery cast to outer join bug (#7090)

when left table column is null and result of subquery is empty, case when will hit null, not false

```
MySQL w2> select t0_27.c_0_2, ((t0_27.c_0_2) IN ((SELECT t0_27.c_0_2 FROM t0 AS t0_27 WHERE (t1_28.c_1_0) IN ('1970-01-18 15:14:39') ) ) ) from t0 AS t0_27, t1 AS t1_28;
+--------------------+-------------------------------------------------------------------------------------+
| c_0_2              | c_0_2 IN (((SELECT c_0_2 FROM t0 AS t0_27 WHERE c_1_0 IN ('1970-01-18 15:14:39')))) |
+--------------------+-------------------------------------------------------------------------------------+
| 0.9736537942295025 | 0                                                                                   |
| 0.5452490003832424 | 0                                                                                   |
| 0.5452490003832424 | 0                                                                                   |
| <null>             | <null>                                                                              |
| <null>             | <null>                                                                              |
| 0.9736537942295025 | 0                                                                                   |
| 0.5452490003832424 | 0                                                                                   |
| 0.5452490003832424 | 0                                                                                   |
+--------------------+-------------------------------------------------------------------------------------+
8 rows in set
Time: 0.077s
```

should be 
```
MySQL w2> select t0_27.c_0_2, ((t0_27.c_0_2) IN ((SELECT t0_27.c_0_2 FROM t0 AS t0_27 WHERE (t1_28.c_1_0) IN ('1970-01-18 15:14:39') ) ) ) from t0 AS t0_27, t1 AS t1_28;
+--------------------+-------------------------------------------------------------------------------------+
| c_0_2              | c_0_2 IN (((SELECT c_0_2 FROM t0 AS t0_27 WHERE c_1_0 IN ('1970-01-18 15:14:39')))) |
+--------------------+-------------------------------------------------------------------------------------+
| 0.9736537942295025 | 0                                                                                   |
| 0.5452490003832424 | 0                                                                                   |
| 0.5452490003832424 | 0                                                                                   |
| <null>             | 0                                                                                   |
| <null>             | 0                                                                                   |
| 0.9736537942295025 | 0                                                                                   |
| 0.5452490003832424 | 0                                                                                   |
| 0.5452490003832424 | 0                                                                                   |
+--------------------+-------------------------------------------------------------------------------------+
8 rows in set
Time: 0.077s

```

wyb committed:
[Lake] Create lake tablet in BE (#7104)

* [Lake] Create lake tablet in BE

1. generate lake TabletMetadataPB and write to object storage.
2. refactor convert thrift schema to protobuf schema.

RowenWoo committed:
[StarRocks On ES] fix wanonly value error when read field (#7103)

sduzh committed:
Fix compilation error (#7108)

sduzh committed:
Add UT for lake::TabletWriter (#7106)

stdpain committed:
[Bug] fix Java UDF crash (#7110)

murphyatwork committed:
Fix the default encoding of largeint (#7113)

liuyehcf committed:
[Enhancement] TopDownRewriteTask add trace log (#7077)

decster committed:
Revert "[Enhancement] Report tablet version after publish version task finish (#6537)" (#7100)

This reverts commit deee5e524384200afc4aa0eaffb3b70bc19a745d.

harui7890 committed:
[Lake] Add starlet group assigner (#7089)

kateshaowanjou committed:
Update README (#7125)

waittttting committed:
Remove the helpers from bdb ReplicationGroupAdmin when dropped fe (#6773)

Remove the helpers from bdb ReplicationGroupAdmin when dropped fe

meegoo committed:
[BugFix] Fix TabletQuorumFailedExceptio msg (#7087)

kangkaisen committed:
[Refactor] Remove old Analyzer from query stmt (#6874)

zhuxt2015 committed:
Merge branch 'main' of https://github.com/StarRocks/starrocks into alter_mv

# Conflicts:
#	fe/fe-core/src/main/java/com/starrocks/persist/EditLog.java
#	fe/fe-core/src/main/java/com/starrocks/persist/OperationType.java
#	fe/fe-core/src/main/java/com/starrocks/sql/StatementPlanner.java
#	fe/fe-core/src/main/java/com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java

sduzh committed:
Typo: reach -> reached (#7132)

murphyatwork committed:
[Enhance] try to release chunks of operator when close (#7086)

decster committed:
[Enhancement] Report tablet version after publish version task finish (#7123)

Fixes #6391

Problem Summary(Required) ：
Currently, BE's publish_version task only reports whether it is successful or not but does not contain which version the replica has and which versions are missing. This will limit FE's ability to publish txn or schedule clone tasks.

For example, if a replica has version 1 2 4 5, and publish_version 3 succeeds, the version should grow to 5, and txns waiting for version 4, 5 to be published can proceed.

This PR also does a major refactor to the PublishVersionTask execution-related code.

zombee0 committed:
[BugFix] estimate output row bytes of OlapScanNode error (#7141)

stdpain committed:
[Enhancement] make log clearer when create Expr (#7133)

Youngwb committed:
Fix project node map error when group by constant (#7135)

Seaven committed:
[Enhancement] Remove decimal column range extract limit (#6997)

murphyatwork committed:
[Enhance] change strategy of pipeline_dop (#6950)

ZiheLiu committed:
[Enhancement] replace query context expired timeout with query_delivery_timeout (#7085)

zhuxt2015 committed:
fix

zhuxt2015 committed:
fix

zhuxt2015 committed:
Merge branch 'main' of https://github.com/StarRocks/starrocks into alter_mv

wanpengfei-git commented:
#### [FE PR Coverage check]

:disappointed: **fail** : 132 / 221 (59.73%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/connector/ConnectorMetadata.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|6|00.00%|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|2|6|33.33%|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|4|11|36.36%|
|:large_blue_circle:|com/starrocks/persist/RenameMaterializedViewLog.java|6|15|40.00%|
|:large_blue_circle:|com/starrocks/persist/ChangeMaterializedViewRefreshSchemeLog.java|6|15|40.00%|
|:large_blue_circle:|com/starrocks/qe/DdlExecutor.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/alter/Alter.java|39|74|52.70%|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|7|12|58.33%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|2|3|66.67%|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|39|44|88.64%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|13|14|92.86%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AsyncRefreshSchemeDesc.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AlterMaterializedViewStatement.java|9|9|100.00%|

zhuxt2015 committed:
fix

wanpengfei-git commented:
#### [FE PR Coverage check]

:disappointed: **fail** : 121 / 211 (57.35%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AsyncRefreshSchemeDesc.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|0|4|00.00%|
|:large_blue_circle:|com/starrocks/connector/ConnectorMetadata.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|6|00.00%|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|2|6|33.33%|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|4|11|36.36%|
|:large_blue_circle:|com/starrocks/persist/RenameMaterializedViewLog.java|6|15|40.00%|
|:large_blue_circle:|com/starrocks/persist/ChangeMaterializedViewRefreshSchemeLog.java|6|15|40.00%|
|:large_blue_circle:|com/starrocks/qe/DdlExecutor.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/alter/Alter.java|39|74|52.70%|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|7|12|58.33%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|2|3|66.67%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|13|16|81.25%|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|30|32|93.75%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AlterMaterializedViewStatement.java|9|9|100.00%|

zhuxt2015 committed:
fix

wanpengfei-git commented:
#### [FE PR Coverage check]

:disappointed: **fail** : 124 / 211 (58.77%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/AstVisitor.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AsyncRefreshSchemeDesc.java|0|2|00.00%|
|:large_blue_circle:|com/starrocks/connector/ConnectorMetadata.java|0|1|00.00%|
|:large_blue_circle:|com/starrocks/journal/JournalEntity.java|0|6|00.00%|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|2|6|33.33%|
|:large_blue_circle:|com/starrocks/persist/EditLog.java|4|11|36.36%|
|:large_blue_circle:|com/starrocks/persist/RenameMaterializedViewLog.java|6|15|40.00%|
|:large_blue_circle:|com/starrocks/persist/ChangeMaterializedViewRefreshSchemeLog.java|6|15|40.00%|
|:large_blue_circle:|com/starrocks/qe/DdlExecutor.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/alter/Alter.java|39|74|52.70%|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|7|12|58.33%|
|:large_blue_circle:|com/starrocks/server/LocalMetastore.java|2|3|66.67%|
|:large_blue_circle:|com/starrocks/sql/analyzer/PrivilegeChecker.java|3|4|75.00%|
|:large_blue_circle:|com/starrocks/sql/parser/AstBuilder.java|13|16|81.25%|
|:large_blue_circle:|com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java|30|32|93.75%|
|:large_blue_circle:|com/starrocks/sql/analyzer/Analyzer.java|2|2|100.00%|
|:large_blue_circle:|com/starrocks/sql/ast/AlterMaterializedViewStatement.java|9|9|100.00%|

ABingHuang commented on fe/fe-core/src/main/java/com/starrocks/alter/Alter.java:
```diff
@@ -189,6 +199,117 @@ public void processDropMaterializedView(DropMaterializedViewStmt stmt) throws Dd
         }
     }
+    public void processAlterMaterializedView(AlterMaterializedViewStatement stmt)
+            throws DdlException, MetaNotFoundException, AnalysisException {
```
AnalysisException seems useless

ABingHuang commented on fe/fe-core/src/main/java/com/starrocks/alter/Alter.java:
```diff
@@ -189,6 +199,117 @@ public void processDropMaterializedView(DropMaterializedViewStmt stmt) throws Dd
+            }
+
+            if (materializedView == null) {
+                throw new MetaNotFoundException("Materialized view " + mvName + " is not find");
```
```suggestion
                throw new MetaNotFoundException("Materialized view " + mvName + " is not found");
```

ABingHuang commented on fe/fe-core/src/main/java/com/starrocks/alter/Alter.java:
```diff
@@ -189,6 +199,117 @@ public void processDropMaterializedView(DropMaterializedViewStmt stmt) throws Dd
+    }
+
+    private void processChangeRefreshScheme(RefreshSchemeDesc refreshSchemeDesc, MaterializedView materializedView)
+            throws DdlException {
```
DdlException seems useless

ABingHuang commented on fe/fe-core/src/main/java/com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java:
```diff
@@ -304,6 +320,43 @@ public Void visitDropMaterializedViewStatement(DropMaterializedViewStmt stmt, Co
+                        try {
+                            TimeUnit.valueOf(unit);
+                        } catch (IllegalArgumentException e) {
+                            throw new SemanticException("Unsupported interval expr: %s", unit);
```
is client can see this exception? if so, add valid unit to log

HangyuanLiu commented on fe/fe-core/src/main/java/com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java:
```diff
@@ -54,6 +61,15 @@ public static void analyze(StatementBase stmt, ConnectContext session) {
 
     static class MaterializedViewAnalyzerVisitor extends AstVisitor<Void, ConnectContext> {
 
+        public enum TimeUnit {
```
TimeUnit does not need to exist

zhuxt2015 replied:
use to check refresh interval time unit

HangyuanLiu commented:
If IntervalLiteral participates in serialization, does it also need to implement SerializedName?
